router/http.go

```diff
@@ -422,109 +384,269 @@ func (s *httpService) getBackendSticky(req *http.Request) (*httputil.ClientConn,
+	for i := range ss {
+		if ss[i] == s {
+			ss[0], ss[i] = ss[i], ss[0]
 		}
 	}
+}
+
+var transport http.RoundTripper = &http.Transport{
+	Dial:                customDial,
+	TLSHandshakeTimeout: 10 * time.Second, // unused, but safer to leave default in place
+}
+
+var dialer = &net.Dialer{
+	Timeout:   1 * time.Second,
+	KeepAlive: 30 * time.Second,
+}
+
+func customDial(network, addr string) (net.Conn, error) {
+	conn, err := dialer.Dial(network, addr)
 	if err != nil {
-		if err != io.EOF && err != httputil.ErrPersistEOF {
-			log.Println("server read err:", err)
-			// TODO: log error
-			fail(sc, req, 502, "Bad Gateway")
+		return nil, dialErr{err}
+	}
+	return conn, nil
+}
+
+type dialErr struct {
+	error
+}
+
+func (s *httpService) forwardAndProxyTCP(w http.ResponseWriter, req *http.Request, addrs []string, stickyAddr string, isSticky bool) {
+	var (
+		backend string
+		err     error
+		upconn  net.Conn
+	)
+	for _, backend = range addrs {
+		req.URL.Host = backend
+		upconn, err = dialer.Dial("tcp", req.URL.Host)
+		if err != nil {
+			// retry, maybe log a message about it
+			continue
 		}
+		defer upconn.Close()
+		break
+	}
+	if upconn == nil {
+		log.Println("no backends available")
+		failAndClose(w, 503)
 		return
 	}
 
-	// TODO: Proxy HTTP CONNECT? (example: Go RPC over HTTP)
-	if res.StatusCode == http.StatusSwitchingProtocols {
-		serverW, serverR := backend.Hijack()
-		clientW, clientR := sc.Hijack()
-		defer serverW.Close()
-		done := make(chan struct{})
-		go func() {
-			serverR.WriteTo(clientW)
-			if cw, ok := clientW.(writeCloser); ok {
-				cw.CloseWrite()
-			}
-			close(done)
-		}()
-		clientR.WriteTo(serverW)
-		serverW.(writeCloser).CloseWrite()
-		<-done
-		return true
+	err = req.Write(upconn)
+	if err != nil {
+		log.Println("error copying request to target:", err)
+		failAndClose(w, 503)
+		return
+	}
+
+	// Need to complete the handshake and set sticky cookie on response, otherwise
+	// websocket reconnections won't go to the right backend
+	upconnbr := bufio.NewReader(upconn)
+	res, err := http.ReadResponse(upconnbr, req)
+	if err != nil {
+		log.Println("http: proxy error:", err)
+		failAndClose(w, 503)
+		return
+	}
+	defer res.Body.Close()
+
+	// Copy the response headers and body over to the downstream ResponseWriter,
+	// the same as done in the non-TCP path.
+	copyHeader(w.Header(), res.Header)
+
+	if res.StatusCode != 101 {
+		// Upgrade was not successful, not going to reuse this connection.
+		w.Header().Set("Connection", "close")
+	}
+
+	if isSticky && stickyAddr != backend {
+		http.SetCookie(w, s.newStickyCookie(backend))
+	}
+
+	w.WriteHeader(res.StatusCode)
+	_, err = io.Copy(w, res.Body) // TODO(bgentry): consider using a flush interval
+	if err != nil {
+		log.Println("reverse proxy copy err:", err)
+		return
+	}
+
+	if res.StatusCode != 101 {
+		return
+	}
+	res.Body.Close() // close this now since we've copied everything
+
+	hj, ok := w.(http.Hijacker)
+	if !ok {
```

Comment:
Can just type assert without checking: `w.(http.Hijacker).Hijack()` which will have an even better error message in the panic.